### PR TITLE
Repaired Hoppers Treating Chests as two entities

### LINKED ISSUE #5088: 2 hoppers attached to double chest treat it as individual chests
Client version: 1.12.2
Server OS: Arch Linux
Cuberite Commit id: 53ae358d8c5db448e7f4b6deac66b3d9e42e59ee

### Expected behavior
Hoppers should treat double chests as double chests, not individual chests (shift clicking items works fine, I guess the client handles that?)

### Actual behavior
Hoppers treat double chests as 2 individual chests, as if the double chest were divided horizontally along the middle when open.

### Steps to reproduce the behavior
Build the following:

- Place (eg.) stone anywhere in the top half of the top chest and wood in the bottom half.
- The stone will be in column 1 line 1 of the output chest and the wood will be in column 1 line 4 of the output chest.

### Server log
```
N/A
```

### PR DESCRIPTION
**Pull request by npresley0506**
Hoppers used to treat double chests as two separate entities and would only pull from one side and deposit in one side. This has been repaired and the hopper now treats the chests as a solid entity. However, this uncovered an issue in that the hopper does not place items on already existing stacks, but will always place them in the first open spot of the chest. This is most likely due to an issue with MoveItemsToGrid and I will open a bug request for it if I can and try to work on it.

Please let me know if you run into any issues with the repairs and I'll do my best to fix them. Any feedback is welcome!

Fixes #5088

### CODE REVIEW

File: src/BlockEntities/HopperEntity.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include "../UI/HopperWindow.h"
 #include "ChestEntity.h"
 #include "FurnaceEntity.h"
+#include <iostream>
```
Comment (12xx12) [inline src/BlockEntities/HopperEntity.cpp:15]: There is something left

File: src/BlockEntities/HopperEntity.cpp [hunk 2]
```diff
@@ -376,30 +377,32 @@ bool cHopperEntity::MoveItemsOut(cChunk & a_Chunk, const cTickTimeLong a_Current
...
-		{ 1, 1,  0},
-		{-1, 1,  0},
-		{ 0, 1,  1},
-		{ 0, 1, -1},
-	} ;
-	for (const auto & ofs: neighborOfs)
+		{1, 1, 0},
+		{-1, 1, 0},
+		{0, 1, 1},
+		{0, 1, -1},
```
Comment (12xx12) [inline src/BlockEntities/HopperEntity.cpp:398]:
```suggestion
		{ 1, 1,  0},
		{-1, 1,  0},
		{ 0, 1,  1},
		{ 0, 1, -1},
```

The spaces improved readability. Same as below

File: src/BlockEntities/HopperEntity.cpp [hunk 3]
```diff
@@ -575,17 +648,45 @@ bool cHopperEntity::MoveItemsToChest(cChunk & a_Chunk, Vector3i a_Coords)
 			// Not the same kind of chest
 			continue;
 		}
-
-		auto chest = static_cast<cChestEntity *>(neighbor->GetBlockEntity(a_Coords + ofs));
+		second = true;
+		auto chest = static_cast<cChestEntity *>(
```
Comment (12xx12) [inline src/BlockEntities/HopperEntity.cpp:652]:
```suggestion
		auto SideChest = static_cast<cChestEntity *>(
```

Maybe rename the variable for better readabilty. This is a bit confusing

File: src/BlockEntities/HopperEntity.cpp [hunk 4]
```diff
@@ -575,17 +648,45 @@ bool cHopperEntity::MoveItemsToChest(cChunk & a_Chunk, Vector3i a_Coords)
...
 		{
-			FLOGWARNING("{0}: A chest entity was not found where expected, at {1} ({2}, {3}})", __FUNCTION__, a_Coords + ofs, ofs.x, ofs.z);
+			FLOGWARNING(
+				"{0}: A chest entity was not found where expected, at {1} "
+				"({2}, {3}})",
+				__FUNCTION__, a_Coords + ofs, ofs.x, ofs.z);
 			continue;
 		}
-		return MoveItemsToGrid(*chest);
+		auto chestAbsCoords = neighbor->RelativeToAbsolute(otherHalfRelCoord);
```
Comment (12xx12) [inline src/BlockEntities/HopperEntity.cpp:662]:
```suggestion
		auto OtherHalfAbsCoords = neighbor->RelativeToAbsolute(OtherHalfRelCoord);
```

**Review by 12xx12 — CHANGES_REQUESTED** [review]
Gameplay is good.

Code seems fine 👍 

Mostly style stuff:
- Your IDE seems to change line ends please undo those (see line 380, 385,...)
- We changed the style of local variables to be capitalized. On some old code this is not the case and we try to upgrade it piece by piece. This is such a case could you do this please?

**Comment by 12xx12** [comment]
And please update the branch to master

File: src/BlockEntities/HopperEntity.cpp [hunk 5]
```diff
@@ -413,21 +416,89 @@ bool cHopperEntity::MoveItemsFromChest(cChunk & a_Chunk)
...
+					{
+						return true;
+					}
+					if (MoveItemsFromGrid(*sideChest))
+					{
+						return true;
+					}
+				}
+				return false;
 			}
```
Comment (bearbin) [inline src/BlockEntities/HopperEntity.cpp:489]: The check for x versus z equivalency is pointless here and just makes the code longer - if you change the inner if condition to be `if ((neighborAbsCoord.x > chestPos.x) || (neighborAbsCoord.z > chestPos.z))` you could then remove the outer if and save half the code.

**Comment by 12xx12** [comment]
Oh - I just noticed you are new to the project. Hi and welcome.
Please add yourself to the CONTRIBUTORS file due to copirght reasons. (and to be included to the newsletter :D)

**Review by bearbin — CHANGES_REQUESTED** [review]
Hi @npresley0506, thanks for taking the time to open this PR and fix the issue. 

Further to @12xx12's comment, you need to revert all of the unrelated formatting changes so this PR is just the code that you are actually changing here. Other than a bit of potential for simplification, the business logic looks fine so once that's done this could be merged in.

You also need to add your name to the CONTRIBUTORS file to indicate that you accept releasing your code under the correct license.

**Comment by tigerw** [comment]
Haven't looked at the code but cChestEntity has a m_Neighbour field, representing the other half, if that helps.

**Comment by 12xx12** [comment]
Why do it easy if there is a more complex solution 😄

**Comment by npresley0506** [comment]
Thanks for the feedback everyone, I think I fixed the style issues and I condensed the logic from the MoveItemsFromChest. That's what I get for coding so late.

I think I managed to get everything, and stopped my IDE from being weird about lines, but if there's anything I missed let me know.

File: src/BlockEntities/HopperEntity.cpp [hunk 6]
```diff
@@ -539,53 +573,78 @@ bool cHopperEntity::MoveItemsFromSlot(cBlockEntityWithItems & a_Entity, int a_Sl
...
 		}
-
-		auto chest = static_cast<cChestEntity *>(neighbor->GetBlockEntity(a_Coords + ofs));
-		if (chest == nullptr)
+		Second = true;
+		auto SideChest = static_cast<cChestEntity *>(Neighbor->GetBlockEntity(a_Coords + ofs));
+		if (SideChest == nullptr)
 		{
-			FLOGWARNING("{0}: A chest entity was not found where expected, at {1} ({2}, {3}})", __FUNCTION__, a_Coords + ofs, ofs.x, ofs.z);
+			FLOGWARNING("{0}: A chest entity was not found where expected, at {1} " "({2}, {3}})", __FUNCTION__, a_Coords + ofs, ofs.x, ofs.z);
```
Comment (12xx12) [inline src/BlockEntities/HopperEntity.cpp:614]:
```suggestion
			FLOGWARNING("{0}: A chest entity was not found where expected, at {1} ({2}, {3}})", __FUNCTION__, a_Coords + ofs, ofs.x, ofs.z);
```

**Review by bearbin — COMMENTED** [review]
Looks fine now, I'll let the other two reviewers have a look to make sure their issues were addressed :)

**Review by 12xx12 — COMMENTED** [review]
One teeeny tiny thing

**Comment by 12xx12** [comment]
That `m_Neighbour` thing is not adressed. This will improve performance (I think)

**Comment by npresley0506** [comment]
m_Neighbour is a private variable of cChestEntity, I'd have to make a function to get to that if that's alright?

**Comment by bearbin** [comment]
Yes, adding a getter function should be fine. Just make sure it's not exported to lua.

**Comment by npresley0506** [comment]
Third times the charm

File: src/BlockEntities/ChestEntity.h [hunk 7]
```diff
@@ -46,6 +46,9 @@ class cChestEntity :
 	/** Search horizontally adjacent blocks for neighbouring chests of the same type and links them together. */
 	void ScanNeighbours();
 
+	/** Returns the value of m_Neighbour. */
+	cChestEntity * ReturnNeighbour();
```
Comment (bearbin) [inline src/BlockEntities/ChestEntity.h:50]: This should be GetNeighbour (but +1 for the proper spelling of neightbour :))

**Comment by npresley0506** [comment]
my bad, should be good now, thanks for the patience haha

**Review by bearbin — APPROVED** [review]
No, thank you for your persistence :)

**Review by 12xx12 — APPROVED** [review]

**Pull request merged**